compendium/DeclarativeServices/src/manager/ComponentManagerImpl.cpp

```diff
@@ -158,25 +211,64 @@ namespace cppmicroservices
             auto logger = GetLogger();
             auto configNotifier = GetConfigNotifier();
 
-            using ActualTask = std::packaged_task<void(std::shared_ptr<CMEnabledState>)>;
+            using ActualTask = std::packaged_task<void(std::shared_ptr<CMEnabledState>, bool)>;
             using PostTask = std::packaged_task<void()>;
 
-            ActualTask task([metadata, bundle, reg, logger, configNotifier](
-                                std::shared_ptr<CMEnabledState> eState) mutable
-                            { eState->CreateConfigurations(metadata, bundle, reg, logger, configNotifier); });
+            ActualTask task(
+                [metadata, bundle, reg, logger, configNotifier, asyncStarted](std::shared_ptr<CMEnabledState> eState,
+                                                                              bool checkExecuted) mutable
+                {
+                    // if this task is being run on spawned thread (not waiting thread), check execution status
+                    if (checkExecuted)
+                    {
+                        bool expected = false;
+                        bool desired = true;
+                        // if asyncStarted is non null AND *asyncStarted==true
+                        if (asyncStarted && !std::atomic_compare_exchange_strong(&(*asyncStarted), &expected, desired))
+                        {
+                            // this is blocking and it has started
+                            return;
+                        }
+                        // else it is non-blocking or it has not started and we now own it
+                    }
+                    // do the task
+                    eState->CreateConfigurations(metadata, bundle, reg, logger, configNotifier);
+                });
 
             std::shared_ptr<CMEnabledState> enabledState = std::make_shared<CMEnabledState>(task.get_future().share());
 
-            PostTask post_task([enabledState, taskPtr = std::make_shared<ActualTask>(std::move(task))]() mutable
-                               { (*taskPtr)(enabledState); });
+            auto taskPtr_ = std::make_shared<ActualTask>(std::move(task));
+
+            // if blocking, cache task and enabledState
+            if (asyncStarted)
+            {
+                asyncTaskMap[asyncStarted] = taskPtr_;
+                asyncTaskStateMap[asyncStarted] = enabledState;
+            }
+            PostTask post_task(
+                [enabledState, taskPtr = taskPtr_]() mutable
+                {
+                    try
+                    {
+                        (*taskPtr)(enabledState, true);
+                    }
+                    catch (...)
+                    {
+                        /*
+                         * task was already executed, by WaitForFuture, calling it
+                         * again will throw. This isexpected, we can just catch and continue
+                         */
+                    }
```

Comment:
I don't think the try catch is necessary as the post task will just return. I just didn't like having an uncaught exception that is likely to happen somewhat regularly in the task even if it won't end up throwing.

Reply:
That being said, I am definitely open to removing it if people believe that that is a fine thing to do

Reply:
Perhaps just remove / replace the comment in the catch block? :smiley:

Reply:
What do you mean? I added the comment to clarify why there is an empty catch block. Do you think the comment should go away but leave the try/catch

Reply:
I don’t think the comment is accurate - it suggests that calling the task will throw if it’s already been executed, but this code should never be reached if the task has already started execution in the `AsyncWorkService`, thanks to the atomic? I think the only reason the task could throw is if the work itself throws, but `std::packaged_task` should be catching any exceptions there to set on the `promise`? So I’m not sure the catch block can ever be hit and it’ll just be reducing code coverage metrics? Perhaps I’ve missed something though?

Reply:
So what I observed is this:

The `taskPtr` here is the function created above by `ActualTask task(...`. That is the function where the atomic checking happens, so this task here (`post_task`) has no knowledge of the atomic (which is inside `taskPtr`) and that is what gets put on the threadpool. So once this thread (the one that was previously waiting) gets allocated, `task` (and then `taskPtr`) will be run. However, `packaged_task` objects have memory in the sense that they know when they have been executed previously and if they have, they [throw ](https://en.cppreference.com/w/cpp/thread/packaged_task/operator()#:~:text=(none)-,Exceptions,-std%3A%3Afuture_error%20on). 

We still need the atomic logic inside this though because the task, allocated to the threadpool, if it wins (as opposed to `WaitForFuture`), needs to set the atomic for `WaitForFuture` to check etc. But if `WaitForFuture` wins, then the atomic logic inside the `task` is never hit because invoking the `taskPtr` throws. 


Let me know if this makes sense

Reply:
Ah, the throwing behaviour of `packaged_task` was what I was missing! Thanks for explaining 😄